Fix IPv6 TCP MSS issue

wirehell opened the pull request:
- Add an MTU check to loopback
- Fixing https://github.com/zephyrproject-rtos/zephyr/issues/56657
- Align IPv4 and IPv6 MSS calculation

rlubos commented on tests/net/socket/tcp/src/main.c:
```diff
@@ -335,18 +335,37 @@ void tcp_server_block_thread(void *vps_sock, void *unused2, void *unused3)
 }
 
-void test_v4_send_recv_large_common(int tcp_nodelay)
+void test_send_recv_large_common(int tcp_nodelay, int family)
```
Please split the test extension into a separate commit.

rlubos commented on tests/net/socket/tcp/src/main.c:
```diff
@@ -416,23 +435,43 @@ static void restore_packet_loss_ratio(void)
+	test_send_recv_large_common(1, AF_INET6);
+	restore_packet_loss_ratio();
+}
+
```
Unneeded empty line.

rlubos reviewed (commented):
Thanks @wirehell for contribution.

It seems that due to extra check in loopback some other test suites need MTU adjustement as well.

Regarding commit messages, the titles need to be prefixed (`drivers: net: loopback:` , `net: tcp:` `tests: net`), also please add some body to the commit messages (https://docs.zephyrproject.org/latest/contribute/guidelines.html#commit-message-guidelines)

wirehell commented:
> It seems that due to extra check in loopback some other test suites need MTU adjustement as well.

I think the loopback MTU check exposes another issue. Tried to dig into it and for the failing test: 
https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/net/socket/udp/src/main.c#L1269
No fragmentation is performed, or at least not using the MTU of the loopback.

Also stumbled upon this behavior which is most likely erroneous if both IPV6 and IPV4 is configured (it doesn't look at packet options, just config). There are no options here though, so this issue doesn't impact here.
https://github.com/zephyrproject-rtos/zephyr/blob/a2f203dd1a7a5110a3339da81e72da024be05ab5/include/zephyr/net/net_pkt.h#L691

I'll fix the things you commented on and will omit the loopback mtu check from the PR.
If I have time I might dig into the above two issues, or open issues.

wirehell committed:
tests: net: Add tests for large IPv6 TCP messages

Add tests for lage IPv6 messages, reusing similar IPv4 test structure.

Signed-off-by: Kenny Johansson <wirehell@gmail.com>

wirehell committed:
tests: net: Set explicit loopback MTU for TCP tests

Set explicit loopback MTU for TCP tests for clarity,
 and to avoid breaking tests if changed elsewhere.

Signed-off-by: Kenny Johansson <wirehell@gmail.com>

wirehell committed:
net: tcp: Fix IPv6 TCP MSS issue

Fixes #56657, incorrect TCP MSS calculation for IPv6.

Previously the supported MSS would incorrectly returned as the MTU if it
was less than the MTU, probably to account for the case where
the network interface MTU check returns 0.

New behaviour is to return the supported MSS as MTU minus header length,
using default MTU for this calculation if network interface MTU check
returns 0.

Signed-off-by: Kenny Johansson <wirehell@gmail.com>

rlubos commented:
> No fragmentation is performed, or at least not using the MTU of the loopback.

IP fragmentation is disabled by default, so no fragmentation is expected. However, I'd rather expect the socket send call to fail if the datagram cannot fit into a single frame.

rlubos reviewed (dismissed):
Looks reasonable.

rlubos commented:
@wirehell Please check the compliance failures.

jukkar commented on subsys/net/ip/tcp.c:
```diff
@@ -3345,13 +3345,20 @@ uint16_t net_tcp_get_supported_mss(const struct tcp *conn)
 	if (family == AF_INET) {
```
In commit "net: tcp: Fix IPv6 TCP MSS issue" could you add more information what/why is being fixed to the commit message. This way user does not necessarily go to github to see the explanation.
We can still have the pointer to the issue, so you could add
"Fixes: #56657"
to the commit message too.

wirehell commented:
> > No fragmentation is performed, or at least not using the MTU of the loopback.
> 
> IP fragmentation is disabled by default, so no fragmentation is expected. However, I'd rather expect the socket send call to fail if the datagram cannot fit into a single frame.

The test that was failing when I tried to enable MTU limit on loopback actually have fragmentation enabled: 
https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/net/socket/udp/testcase.yaml#L15
I tried to trace a bit but couldn't see any fragmentation being done except on the receiving end. But I didn't look that hard as I don't have any use for that feature.
